pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"path/filepath"
+	"reflect"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+
+	cpapp "github.com/clusterlink-net/clusterlink/cmd/cl-controlplane/app"
+	dpapp "github.com/clusterlink-net/clusterlink/cmd/cl-dataplane/app"
+	api "github.com/clusterlink-net/clusterlink/pkg/apis/clusterlink.net/v1alpha1"
+	cpapi "github.com/clusterlink-net/clusterlink/pkg/controlplane/api"
+	dpapi "github.com/clusterlink-net/clusterlink/pkg/dataplane/api"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	ControlPlaneName      = "cl-controlplane"
+	DataPlaneName         = "cl-dataplane"
+	GoDataPlaneName       = "cl-go-dataplane"
+	IngressName           = "cl-external-svc"
+	OperatorNameSpace     = "clusterlink-operator"
```

Comment:
You should be able to read this from an environment variable and not hard code it.
```go
// Namespace returns the Pod namespace, either from environment or k8s files
func Namespace() (string, error) {
	const nsFile = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
	if ns := os.Getenv("POD_NAMESPACE"); ns != "" {
		return ns, nil
	}
	if data, err := ioutil.ReadFile(nsFile); err == nil {
		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
			return ns, nil
		}
		return "", err
	}
	return "", nil
}
```

Reply:
I will remove it when we have the cl-controller (can create import service in different namespaces)
I will add TODO.